### PR TITLE
4.x native datetime

### PR DESCRIPTION
**Pull request by ADmad**
As discussed in #12820, this PR updates `FormHelper` to generate native HTML inputs for all available datetime types. I have modified `DateTimeWidget` accordingly and added new `YearWidget` to generate selectbox for years.

Still have to do lot of cleanup.

What should be done for the individual `FormHelper()` methods for generating single selects like `day()`, `hour()`, `minute()`, `seconds()`, `meridian()`?  My vote is to just drop them. They can still be generated using `select()` with appropriate options list if some needs them.

### CODE REVIEW

**Comment by dereuromark** [comment]
We should make sure that all the "old" ways have an easy upgrade path to an existing method somewhere.
That eases the migration efforts for existing apps IMO.

File: src/View/Widget/YearWidget.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            $data['end'] = $data['max'];
+        }
+
+        if (!empty($data['val'])) {
+            $data['start'] = min($data['val'], $data['start']);
+            $data['end'] = max($data['val'], $data['end']);
+        }
+
+        if ($data['end'] < $data['start']) {
+            throw new RuntimeException('Max year cannot be less than min year');
```
Comment (markstory) [inline src/View/Widget/YearWidget.php]: Could be a `LogicException` or `InvalidArgumentException` 🏌️

**Review by markstory — COMMENTED** [review]
I second moving the existing widget into a plugin for people who want to upgrade but can't update their UI.

**Comment by ADmad** [comment]
Along with existing DateTimeWidget the shim plugin will also need a FormHelper with retains all the options juggling that is currently done for datetime inputs.

**Comment by markstory** [comment]
> will also need a FormHelper with retains all the options juggling that is currently done for datetime inputs.

Yeah that is going to be awkward 😢

File: src/Database/Type/DateTimeType.php [hunk 2]
```diff
@@ -50,6 +50,7 @@ class DateTimeType extends BaseType
      */
     protected $_format = [
         'Y-m-d H:i:s',
+        'Y-m-d\TH:i:s',
```
Comment (dereuromark) [inline src/Database/Type/DateTimeType.php]: Should we backport this to 3.x?

**Comment by dereuromark** [comment]
All right, we have a branch now: https://github.com/dereuromark/cakephp-shim/tree/4.x#new-shims

**Comment by ADmad** [comment]
Ancillary work of PR to Shim plugin and docs PR is done. This should be good to merge now.

**Pull request merged**